pkg/pool/linked_list.go

```diff
@@ -0,0 +1,100 @@
+package pool
+
+import "sync"
+
+func NewDoubleLinkedList[T any]() *DoubleLinkedList[T] {
+	return &DoubleLinkedList[T]{}
+}
+
+type Node[T any] struct {
+	prev  *Node[T]
+	next  *Node[T]
+	value T
+}
+
+func (n *Node[T]) Value() T {
+	return n.value
+}
+
+type DoubleLinkedList[T any] struct {
+	lock sync.Mutex
+	head *Node[T]
+	len  uint64
+}
+
+func (l *DoubleLinkedList[T]) Len() uint64 {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	return l.len
+}
+
+func (l *DoubleLinkedList[T]) Insert(key T) *Node[T] {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	newNode := &Node[T]{
+		value: key,
+		next:  l.head,
+	}
+
+	if l.head != nil {
+		l.head.prev = newNode
+	}
+	l.head = newNode
+
+	l.len++
+
+	return newNode
+}
+
+func (l *DoubleLinkedList[T]) Delete(node *Node[T]) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	if node == l.head {
+		l.head = node.next
+	}
+
+	if node.next != nil {
+		node.next.prev = node.prev
+	}
+
+	if node.prev != nil {
+		node.prev.next = node.next
+	}
+
+	l.len--
+}
+
+func (l *DoubleLinkedList[T]) Shift() *Node[T] {
```

Comment:
You should add comments to these functions to make it clear what they do, I'm not really sure what this function is supposed to do to be honest.

Also, don't bother allocating a new node here or anything like that - just return the pointer stored in l.head, it's up to the user to not modify anything and it avoids an allocation.

Reply:
I've renamed it to `PopFirst` to be a bit more descriptive, I was following the JS convention for `shift`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/shift

This is used to `get` a connection from the pool, which also removes it.

Resolved in e28b3ab09456db7a76e055e0b46c7d3ceef01589